extension/healthcheckv2extension/internal/status/aggregation.go

```diff
@@ -0,0 +1,152 @@
+}
+
+type ErrorPriority int
+
+const (
+	PriorityPermanent ErrorPriority = iota
+	PriorityRecoverable
+)
+
+type aggregationFunc func(*AggregateStatus) Event
+
+// The purpose of aggregation is to ensure that the most relevant status bubbles
+// upwards in the aggregate status. This aggregation func prioritizes lifecycle
+// events (including FatalError) over PermanentError and RecoverableError
+// events. The priority argument determines the priority of PermanentError
+// events vs RecoverableError events. Lifecycle events will have the timestamp
+// of the most recent event and error events will have the timestamp of the
+// first occurrence.
+func newAggregationFunc(priority ErrorPriority) aggregationFunc {
+	permanentPriorityFunc := func(seen map[component.Status]struct{}) component.Status {
+		if _, isPermanent := seen[component.StatusPermanentError]; isPermanent {
+			return component.StatusPermanentError
+		}
+		if _, isRecoverable := seen[component.StatusRecoverableError]; isRecoverable {
+			return component.StatusRecoverableError
+		}
+		return component.StatusNone
+	}
+
+	recoverablePriorityFunc := func(seen map[component.Status]struct{}) component.Status {
+		if _, isRecoverable := seen[component.StatusRecoverableError]; isRecoverable {
+			return component.StatusRecoverableError
+		}
+		if _, isPermanent := seen[component.StatusPermanentError]; isPermanent {
+			return component.StatusPermanentError
+		}
+		return component.StatusNone
+	}
+
+	errPriorityFunc := permanentPriorityFunc
+	if priority == PriorityRecoverable {
+		errPriorityFunc = recoverablePriorityFunc
+	}
+
+	statusFunc := func(st *AggregateStatus) component.Status {
+		seen := make(map[component.Status]struct{})
+		for _, cs := range st.ComponentStatusMap {
+			seen[cs.Status()] = struct{}{}
+		}
+
+		// All statuses are the same. Note, this will handle StatusOK and StatusStopped as these two
+		// cases require all components be in the same state.
+		if len(seen) == 1 {
+			for st := range seen {
+				return st
+			}
+		}
+
+		// Handle mixed status cases
+		if _, isFatal := seen[component.StatusFatalError]; isFatal {
+			return component.StatusFatalError
+		}
+
+		if _, isStarting := seen[component.StatusStarting]; isStarting {
+			return component.StatusStarting
+		}
+
+		if _, isStopping := seen[component.StatusStopping]; isStopping {
+			return component.StatusStopping
+		}
+
+		if _, isStopped := seen[component.StatusStopped]; isStopped {
+			return component.StatusStopping
+		}
+
+		return errPriorityFunc(seen)
+	}
+
+	return func(st *AggregateStatus) Event {
+		var ev, lastEvent, matchingEvent Event
+		status := statusFunc(st)
+		isError := component.StatusIsError(status)
+
+		for _, cs := range st.ComponentStatusMap {
+			ev = cs.Event
+			if lastEvent == nil || lastEvent.Timestamp().Before(ev.Timestamp()) {
+				lastEvent = ev
+			}
+			if status == ev.Status() {
+				switch {
+				case matchingEvent == nil:
+					matchingEvent = ev
+				case isError:
```

Comment:
can you add some comments/tests for this specifically? I assume this is to return the earliest error, but would be nice to see this described more properly

Reply:
I marked this unresolved because I'm not clear on this either, and I don't believe we have tests for a case where the top-level status is an error and we hit this case. I understand why we do the ordering if we hit this, but it would be nice to see more details about why this case is separate from the one below where we return `matchingEvent` if `isError` is set (on line 139 as of this comment).